### DISCUSSION TITLE
Custom Inpainting Model Trainer with Custom GUI

### DISCUSSION THREAD

**Discussion by Rowan3D** [nagadomi/nunif#737] [Show and tell]
With Claude's help using the desktop app i have created a custom GUI for training inpainting models to work with iw3.
It includes several very helpful features including a data collection step with api support for a couple websites.
A MUCH improved warping system that supports much higher divergence without breaking.
New more powerful inpaint model presets that are 9M 16M 38M parameters for the small medium and large variant over the just over 2M default light inpaint.
Support for training higher res natively.
And a bunch of nice to have features, it is portable and needs the path to a dev version of iw3 to work.
<img width="1913" height="988" alt="Screenshot 2026-09-15 010204" src="https://github.com/user-attachments/assets/fcd730be-9c18-4bab-bf92-c3b12acd9a19" />
<img width="530" height="872" alt="Screenshot 2026-09-15 010229" src="https://github.com/user-attachments/assets/e2272079-0e1e-4c12-b9a3-62c4340c6d90" />
<img width="528" height="977" alt="Screenshot 2026-09-15 010252" src="https://github.com/user-attachments/assets/fe96a4b2-b61e-4a9f-a9c0-0f7e02276547" />
<img width="531" height="821" alt="Screenshot 2026-09-15 010310" src="https://github.com/user-attachments/assets/c1ac182b-1f58-45fa-aa4c-e6b1ef860d49" />
<img width="533" height="486" alt="Screenshot 2026-09-15 010324" src="https://github.com/user-attachments/assets/b390b916-611c-4713-a889-db56298a44b1" />
<img width="529" height="446" alt="Screenshot 2026-09-15 010337" src="https://github.com/user-attachments/assets/77a9cfd0-514f-4d06-bfd8-43d17e9f128a" />
[iw3_Inpaint_Trainer_GUI.zip](https://github.com/user-attachments/files/32217996/iw3_Inpaint_Trainer_GUI.zip)
Its still in WIP and still has a couple quirks but seems to work end to end at the moment so if anyone is interested give it a try!
And if you get a good model out of it be sure to share.

**Comment by QuickscopingFTW** [comment]
Also i changed the optimizer to CAME instead of Adamw as it learns faster, Adamw is still available to use if needed though

**Comment by QuickscopingFTW** [comment]
I am currently starting a training run using the medium custom model at 384 res, i am using 5,000 videos with 350,000 image pairs, it will take around 80 hours to train, i will share updates if it is going well, here is the first eval image at epoch 4:
<img width="1226" height="340" alt="Screenshot 2026-09-15 122101" src="https://github.com/user-attachments/assets/ca9d2999-03a8-4287-91bc-6f9b8f5eb9bd" />

**Reply by Jokeru18** [reply]
That's a huge dataset for a one man show. May I ask where you sourced it?

**Reply by QuickscopingFTW** [reply]
pexels and pixabax, both have a free api key

**Comment by QuickscopingFTW** [comment]
Ok training update:
i am currently on epoch 199/597 and the current model seems to far surpass the base inpaint already, here is an example at VERY high depth at full inpaint res of 1920x1080:

Default light inpaint v1:
<img width="314" height="259" alt="Screenshot 2026-09-16 115119" src="https://github.com/user-attachments/assets/90b8a918-35db-4742-92c4-d5f72e1737fe" />

199 epoch custom medium model:
<img width="316" height="258" alt="Screenshot 2026-09-16 115131" src="https://github.com/user-attachments/assets/c8db08be-3d29-4348-a116-b5a8681b250e" />

You can clearly see on this test footage screenshot it has much better understanding of structure, an is less pixelated overall, in the actual video it seems to have less overall flicker too

**Comment by QuickscopingFTW** [comment]
it currently still falls behind a little in cases of smaller divergence but hopefully it will equal out and be as good as the original at small divergence, at the moment it surpasses it in high divergence only

**Comment by QuickscopingFTW** [comment]
[iw3_Inpaint_Trainer_GUI_V1.01.zip](https://github.com/user-attachments/files/32287420/iw3_Inpaint_Trainer_GUI_V1.01.zip)
Hotfix for the trainer with several fixes:
-Tooltips now stay visible when hovering over them.
-Live view stats are more accurate and have more info.
-Live view image comparison no longer flickers to the chart when switching between epoch comparisons.
-Fixed a major issue with the import system where it would not make a standalone copy of the best checkpoint so if it was overridden it would also be overridden in iw3 automatically, it now has a copy of the model you imported saved and unchanging.
-Several other small fixes.
--INSTALLATION--
If you are already using the tool all you need to do is replace the app folder as well as the readme, even if you have a run going it will be unaffected, make sure to stop the run and close the program before updating.

**Comment by KolaKater** [comment]


**Reply by DepthAuteur** [reply]
... without IW3 means with ...? Just curious about interesting alternatives. If you allow me to guess which MDE you used for those pictures my guess would be Apples Depth Pro :-).

Thanks for answering and sharing.

Best wishes to you @KolaKater.

**Comment by QuickscopingFTW** [comment]
The model im training is specifically a video version, im not too sure if it will be as good on images, my trainer does support image training too though if you would want to train that type, it should be a bit easier to train an image version too

**Comment by QuickscopingFTW** [comment]
<img width="1252" height="809" alt="image" src="https://github.com/user-attachments/assets/3a88bc7c-1719-42f0-86fb-4603f8ee3c81" />
the model is currently just starting the second run of either 3 or 4 that i will do

**Comment by KolaKater** [comment]


**Reply by QuickscopingFTW** [reply]
damn, i have a 4080, its alright for light training but anything like this takes a fair few days still lol

**Comment by QuickscopingFTW** [comment]
Update:
[iw3_Inpaint_Trainer_GUI_V1.10.zip](https://github.com/user-attachments/files/32309329/iw3_Inpaint_Trainer_GUI_V1.10.zip)
new version with a lot of improvements, training should be a lot faster and the whole ui is a lot more robust and responsive and a fair few edge cases have been fixed
-Fixed nunif directory not saving or working if a config file has not been created yet
-Fixed black flashing when switching epoch samples
-Added tf32 speedup to training
-Added Cudnn benchmark to training for speed
-Added ema to training
-Added a discriminator pass on the 3rd round of training to really refine it near the end
-Subfolders of videos now work properly in the downloads folder
-Fixed broken number of images in a folder when pausing in the data prep step
-Changed so the refine warp setting is now on by default
-Some theme color fixes
-Changed several default options for the better settings
-Various other bug fixes and improvements
--INSTALLATION--
This patch will not work on an already training run, you can transfer your data folder to the new one if you are happy with it

**Comment by QuickscopingFTW** [comment]
Update:
[iw3_Inpaint_Trainer_GUI_V1.21.zip](https://github.com/user-attachments/files/32418158/iw3_Inpaint_Trainer_GUI_V1.21.zip)
new version with a big quality boost option and speedup
--Added several discriminator models to run alongside the training, they do a much better job at rewarding the model learning actual fine detail, the old version just saw how close the average color of the mask area was, this new one respects actual detail.
--Added compilation of the model.
--Added compilation of the discriminator.
--Added EMA.
--Several speedup fixes.
--Several fixes.

**Comment by QuickscopingFTW** [comment]
here are a couple examples from my current training run at 502 epochs,, this is using the exact same settings in the iw3 tool, the top images are with light inpaint v1 and the bottom ones are with my custom trained model so far, both done at 16 strength.
<img width="173" height="292" alt="Screenshot 2026-09-19 164359" src="https://github.com/user-attachments/assets/60d65eaa-81de-417e-87c6-1c4d87c0d602" />
<img width="178" height="290" alt="Screenshot 2026-09-19 164425" src="https://github.com/user-attachments/assets/fd20a5d0-fe3f-4a10-b641-5ea05671e4b6" />
<img width="227" height="206" alt="Screenshot 2026-09-19 164527" src="https://github.com/user-attachments/assets/b144aeb1-6677-4546-91c0-3d65647999df" />
<img width="227" height="206" alt="Screenshot 2026-09-19 164540" src="https://github.com/user-attachments/assets/0effa6c8-d5cf-43bc-be57-b1ef8884468d" />
@nagadomi what do you think?

**Reply by QuickscopingFTW** [reply]
i trained this model with actual 16 divergence support out the gate and it seems to be working a treat so far!

**Reply by nagadomi** [reply]
The divergence 5 limit was decided from the beginning, and it is expected that the results are worse beyond that.
It is expected that larger divergence will work better if the range of the training data generation settings is expanded and the model parameters and receptive field are increased. I believe I wrote about these limitations before.

Currently, the inpainting model YAML configuration only allows adding ".pth" files. I will make it possible to automatically load the architecture definition as well. Then models trained with this can also be loaded just by placing the files there.

**Comment by QuickscopingFTW** [comment]
https://pixeldrain.com/u/dMcPVGB1
this link contains Rowan_iw3_High_Depth_inpaint_Installer.zip
Here is my first release of the model, it comes with an auto installer as it changes some of the masking settings, it will not mess with any files in the actual program.
i have finally trained my high depth/disparity inpainting model for iw3, this is an auto installer, either put it in your nunif-windows/nunif directory and launch it or put it elsewhere and when you run it give it the path to the same place.
I might still train the model longer and if i do i will send the model file which you can just add to the installer to install it again.
I recommend edge fix 3-2 and 1-1 inpaint mask dilation, but feel free to experiment!
feedback is welcome!
This was trained up to epoch 594 at 640 res with 2-16 divergence on 175,000 image pairs from 5,000 videos in 12 frame chunks using my custom 16M parameter medium model using the CAME optimizer and discriminator for all 3 199 epoch runs.
This can obviously be improved a fair bit by a larger dataset at higher res for longer with a higher discriminator setting.
But this is what i could do in a reasonable amount of time on my rtx 4080, it took around 5 days straight of training.
@nagadomi i dont know if you would be willing to try use this model but any feedback would be welcome if so!

**Reply by QuickscopingFTW** [reply]
this install also optionally includes my low res inpaint fix so you can run the inpaint at lower res if needed while retaining the res of the original video, i would recommend 1280 inpaint max width for most things, 1920 is fine for landscape usually though.

**Reply by QuickscopingFTW** [reply]
also this was training UP to 16 divergence/strength so going over is untested, but over that is VERY high

**Reply by QuickscopingFTW** [reply]
btw this whole trainer and everything was done heavily with claude's help so i expect there are several things that are not optimal or done properly so i apologize for anything going wrong with training that i didnt test, i did try to find as many edge cases as i could

**Reply by nagadomi** [reply]
I am still away and can only use my phone, so I cannot test it myself. However, since you are using a coding agent correctly, I expect there are not many serious issues.

The problem is that there is currently no easy, safe, and clean way to provide models trained with this to users. Replacing files outside of version control can always conflict with changes I make.
As I mentioned above, there is already a mechanism to load ".pth" files from the YAML configuration, so we need to check whether simply extending that mechanism is enough.

Also if you plan to maintain this, I recommend putting it on your GitHub, or at least using a version control system. That way, if I make a breaking change, you can check the history and adapt to it quickly.

**Reply by QuickscopingFTW** [reply]
i will look into maybe trying to put it on github, never done it before so it will be a first lol, also i do utilize the .yaml that you have to add it in but i also do some other things too

**Comment by QuickscopingFTW** [comment]
[iw3_Inpaint_Trainer_GUI_V1.22.zip](https://github.com/user-attachments/files/32427351/iw3_Inpaint_Trainer_GUI_V1.22.zip)
Here is the most recent version of the trainer, fixed a couple ui bugs and added a feature to overrun the 3 199 epoch run default and go for longer.

**Comment by QuickscopingFTW** [comment]
I am SUPER happy with the results i am getting on the videos i am trying, i mostly made this tool for me as i like high depth and was tired of waiting for someone to make a good inpaint that was quick and high quality, i honestly think this rivals things like m2svid in some aspects while being way smaller and faster.
Even though it was mostly for me i want to make it all available public so other people can enjoy it and use it if they want as well as train their own if they think they can do better with my trainer.

**Comment by QuickscopingFTW** [comment]
https://pixeldrain.com/u/GeiYwQ2q
Rowan_iw3_High_Depth_inpaint_Installer_v1.1
new version with some slight improvements!
thanks to @justcheckking on the community discord it now runs slightly faster and with less vram overhead
-added preserve screen boarder to the install options for forward fill
-fixed window frames being duplicated if using detect scene boundaries causing a noticeable vram spike

**Reply by QuickscopingFTW** [reply]
to install the new version use the uninstall.bat at nunif-windows\nunif\nt_inpaint and then reinstall with this new one

**Comment by QuickscopingFTW** [comment]
https://pixeldrain.com/u/nMcMLUoY
Rowan_iw3_High_Depth_inpaint_Installer_v1.1.1
hotfix patch with several performance improvements and fixes

**Reply by QuickscopingFTW** [reply]
tested results on my rtx 4080:
settings:
16 strength/divergence
Convergence plane sod_v1 - 0.5
both eyes
forward_inpaint
Rowan_High-Depth_Inpaint_e594-Medium
Inpaint overlap frames 3-3
Inpaint mask dilation 1-1
Inpaint max width 1080 (with my low res fix from the installer)
Depth model Any_V3_Mono at 720 res
Edge Fix 3-2
Depth Anti Aliasing On
Flicker reduction 0.99 150
Scene Boundary Detection On

Average on 1920x1080 input video:
3-3.5fps

**Reply by KolaKater** [reply]
could you please post a screenshot of your settings, thank you

**Comment by Rowan3D** [comment]
[iw3_Inpaint_Trainer_GUI_V1.23.zip](https://github.com/user-attachments/files/32471608/iw3_Inpaint_Trainer_GUI_V1.23.zip)
new version of the trainer adds parity between the standalone installer and the training gui so they install the model to iw3 the exact same way and with the optional features i add

**Comment by Rowan3D** [comment]
example of 16 strength on a nightmare scene to infill, top is normal light inpaint, bottom is my current best custom model:
<img width="242" height="188" alt="Screenshot 2026-09-21 144426" src="https://github.com/user-attachments/assets/bfc1288e-7dcf-4b11-ae40-e94a2ac6feaf" />
<img width="242" height="189" alt="Screenshot 2026-09-21 144437" src="https://github.com/user-attachments/assets/533d2acc-3c2b-412c-ba74-c79d853b36dd" />

**Comment by Rowan3D** [comment]
any new stuff will be posted in my other discussion thread
